### PR TITLE
Rework profile area for user and room settings to be more clear

### PR DESCRIPTION
**Pull request by turt2live**
Fixes https://github.com/vector-im/element-web/issues/15108

By design request:
* Fix header spacing on the user settings general tab to match the other tabs
* Implement design for better context interactions, spacing, and alignment

Freehand (**designer - please review**):
* Adhoc build: https://dev.t2host.io/reactsdk5243
* Border on the edit icon when there's a placeholder icon to differentiate them
* Transition effect on hover (went for a function that felt fun to play with)
* The room settings layout

Screenshots (user settings):
image
image

Screenshots (room settings):
image
image

### CODE REVIEW

**Comment by turt2live** [comment]
Design's out of band review:
image
image

#F4F6FA on light
#21262C on dark

File: res/css/views/settings/_AvatarSetting.scss [hunk 1]
```diff
@@ -15,13 +15,55 @@ limitations under the License.
...
-    width: $font-88px;
-    height: $font-88px;
-    margin-left: 13px;
+    width: 90px;
+    height: 90px;
+    margin-top: 8px;
     position: relative;
 
+    .mx_AvatarSetting_hover {
+        transition: opacity 0.08s cubic-bezier(.46, .03, .52, .96); // quadratic
```
Comment (t3chguy) [inline res/css/views/settings/_AvatarSetting.scss:24]: this should go into a variable so we can refer to it for consistent transitions

File: res/css/views/settings/_AvatarSetting.scss [hunk 2]
```diff
@@ -41,8 +83,9 @@ limitations under the License.
     & > img,
     .mx_AvatarSetting_avatarPlaceholder {
         display: block;
-        height: $font-88px;
-        border-radius: 4px;
+        height: 90px;
+        border-radius: 90px;
```
Comment (t3chguy) [inline res/css/views/settings/_AvatarSetting.scss:87]: the radius of the border exceeds half the length of an edge, 50% is clearer

File: res/css/views/settings/_AvatarSetting.scss [hunk 3]
```diff
@@ -58,6 +101,29 @@ limitations under the License.
         left: 0;
         right: 0;
     }
+
+    .mx_AvatarSetting_uploadButton {
+        width: 32px;
+        height: 32px;
+        border-radius: 32px;
```
Comment (t3chguy) [inline res/css/views/settings/_AvatarSetting.scss:108]: same as above

File: res/css/views/settings/tabs/user/_GeneralUserSettingsTab.scss [hunk 4]
```diff
@@ -22,6 +22,13 @@ limitations under the License.
     margin-top: 0;
 }
 
+// TODO: Make this selector less painful
```
Comment (t3chguy) [inline res/css/views/settings/tabs/user/_GeneralUserSettingsTab.scss:25]: might as well do that whilst you're here, doesn't seem large enough to be done on its own

File: src/components/views/room_settings/RoomProfileSettings.js [hunk 5]
```diff
@@ -150,7 +159,11 @@ export default class RoomProfileSettings extends React.Component {
         const AccessibleButton = sdk.getComponent('elements.AccessibleButton');
         const AvatarSetting = sdk.getComponent('settings.AvatarSetting');
         return (
-            <form onSubmit={this._saveProfile} autoComplete="off" noValidate={true}>
+            <form
+                onSubmit={this._saveProfile}
+                autoComplete="off" noValidate={true}
```
Comment (t3chguy) [inline src/components/views/room_settings/RoomProfileSettings.js:164]: one per line

File: src/components/views/room_settings/RoomProfileSettings.js [hunk 6]
```diff
@@ -169,10 +182,20 @@ export default class RoomProfileSettings extends React.Component {
                         uploadAvatar={this.state.canSetAvatar ? this._uploadAvatar : undefined}
                         removeAvatar={this.state.canSetAvatar ? this._removeAvatar : undefined} />
                 </div>
-                <AccessibleButton onClick={this._saveProfile} kind="primary"
-                                  disabled={!this.state.enableProfileSave}>
-                    {_t("Save")}
-                </AccessibleButton>
+                <div className="mx_ProfileSettings_buttons">
+                    <AccessibleButton
+                        onClick={this._clearProfile} kind="link"
```
Comment (t3chguy) [inline src/components/views/room_settings/RoomProfileSettings.js:187]: same as above

File: src/components/views/room_settings/RoomProfileSettings.js [hunk 7]
```diff
@@ -169,10 +182,20 @@ export default class RoomProfileSettings extends React.Component {
...
+                        disabled={!this.state.enableProfileSave}
+                    >
+                        {_t("Cancel")}
+                    </AccessibleButton>
+                    <AccessibleButton
+                        onClick={this._saveProfile} kind="primary"
```
Comment (t3chguy) [inline src/components/views/room_settings/RoomProfileSettings.js:193]: same as above

File: src/components/views/settings/AvatarSetting.js [hunk 8]
```diff
@@ -14,42 +14,46 @@ See the License for the specific language governing permissions and
...
-    const AccessibleButton = sdk.getComponent('elements.AccessibleButton');
-
-    const openImageView = useCallback(() => {
-        const ImageView = sdk.getComponent("elements.ImageView");
-        Modal.createDialog(ImageView, {
-            src: avatarUrl,
-            name: avatarName,
-        }, "mx_Dialog_lightbox");
-    }, [avatarUrl, avatarName]);
+    const [isHovering, setIsHovering] = useState();
```
Comment (t3chguy) [inline src/components/views/settings/AvatarSetting.js:24]: specify initialState

File: src/components/views/settings/ProfileSettings.js [hunk 9]
```diff
@@ -65,6 +65,15 @@ export default class ProfileSettings extends React.Component {
         });
     };
 
+    _clearProfile = async (e) => {
+        e.stopPropagation();
+        e.preventDefault();
+
+        if (!this.state.enableProfileSave) return;
+        this._removeAvatar();
+        this.setState({enableProfileSave: false, displayName: this.state.originalDisplayName});
```
Comment (t3chguy) [inline src/components/views/settings/ProfileSettings.js:74]: can this be split over multiple lines

File: src/components/views/settings/ProfileSettings.js [hunk 10]
```diff
@@ -144,18 +153,26 @@ export default class ProfileSettings extends React.Component {
         const AccessibleButton = sdk.getComponent('elements.AccessibleButton');
         const AvatarSetting = sdk.getComponent('settings.AvatarSetting');
         return (
-            <form onSubmit={this._saveProfile} autoComplete="off" noValidate={true}>
+            <form
+                onSubmit={this._saveProfile}
+                autoComplete="off" noValidate={true}
```
Comment (t3chguy) [inline src/components/views/settings/ProfileSettings.js:158]: one per line

File: src/components/views/settings/ProfileSettings.js [hunk 11]
```diff
@@ -144,18 +153,26 @@ export default class ProfileSettings extends React.Component {
...
+                className="mx_ProfileSettings_profileForm"
+            >
                 <input type="file" ref={this._avatarUpload} className="mx_ProfileSettings_avatarUpload"
                        onChange={this._onAvatarChanged} accept="image/*" />
                 <div className="mx_ProfileSettings_profile">
                     <div className="mx_ProfileSettings_controls">
+                        <span className="mx_SettingsTab_subheading">{_t("Profile")}</span>
+                        <Field
+                            label={_t("Display Name")}
+                            type="text" value={this.state.displayName}
```
Comment (t3chguy) [inline src/components/views/settings/ProfileSettings.js:168]: one per line

File: src/components/views/settings/ProfileSettings.js [hunk 12]
```diff
@@ -164,10 +181,20 @@ export default class ProfileSettings extends React.Component {
                         uploadAvatar={this._uploadAvatar}
                         removeAvatar={this._removeAvatar} />
                 </div>
-                <AccessibleButton onClick={this._saveProfile} kind="primary"
-                                  disabled={!this.state.enableProfileSave}>
-                    {_t("Save")}
-                </AccessibleButton>
+                <div className="mx_ProfileSettings_buttons">
+                    <AccessibleButton
+                        onClick={this._clearProfile} kind="link"
```
Comment (t3chguy) [inline src/components/views/settings/ProfileSettings.js:186]: one per line

File: src/components/views/settings/ProfileSettings.js [hunk 13]
```diff
@@ -164,10 +181,20 @@ export default class ProfileSettings extends React.Component {
...
+                        disabled={!this.state.enableProfileSave}
+                    >
+                        {_t("Cancel")}
+                    </AccessibleButton>
+                    <AccessibleButton
+                        onClick={this._saveProfile} kind="primary"
```
Comment (t3chguy) [inline src/components/views/settings/ProfileSettings.js:192]: one per line

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: res/css/views/settings/_AvatarSetting.scss (hunk 2 above)
Comment (turt2live) [reply res/css/views/settings/_AvatarSetting.scss:87]: The convention has always been to use the length of any side when using a square.

File: res/css/views/settings/tabs/user/_GeneralUserSettingsTab.scss (hunk 4 above)
Comment (turt2live) [reply res/css/views/settings/tabs/user/_GeneralUserSettingsTab.scss:25]: It's not trivial enough to be included here imo

File: src/components/views/room_settings/RoomProfileSettings.js (hunk 5 above)
Comment (turt2live) [reply src/components/views/room_settings/RoomProfileSettings.js:164]: This isn't a rule we have?

File: src/components/views/settings/ProfileSettings.js (hunk 9 above)
Comment (turt2live) [reply src/components/views/settings/ProfileSettings.js:74]: Not clear on why

File: src/components/views/room_settings/RoomProfileSettings.js (hunk 5 above)
Comment (t3chguy) [reply src/components/views/room_settings/RoomProfileSettings.js:164]: Neither is the indent-by-4 vs natural indenting in WebStorm, its a convention

Comment (turt2live) [reply src/components/views/room_settings/RoomProfileSettings.js:164]: We also have a history of grouping similar properties onto a single line.

**Review by t3chguy — APPROVED** [review]

**Review by t3chguy — APPROVED** [review]

**Pull request merged**